File: extensions/syncforreddit/src/main/java/app/morphe/extension/syncforreddit/ui/comments/InlineCommentMediaPatch.java
package app.morphe.extension.syncforreddit.ui.comments;

import java.util.Locale;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.syncforreddit.settings.SyncUpSettings;

/**
 * Whether a link in a comment is drawn as the picture it points at, rather than as a chip naming
 * where it goes.
 *
 * <p>Sync already draws some of them inline, but only where the address ends in one of five
 * extensions or belongs to one of a handful of named services. Anything else — a Reddit gallery
 * link, an Imgur page, a RedGifs or Giphy link, an address that names no extension at all —
 * falls through to the chip, which is what makes the behaviour look arbitrary from a thread
 * where one comment draws and the next does not.
 *
 * <p>This widens that test to everything Sync can actually draw, on a setting of its own.
 *
 * @noinspection unused
 */
public final class InlineCommentMediaPatch {
    private static final String INLINE_EVERYTHING = "sync_up_inline_comment_media";

    /** Hosts that serve a picture or a video whatever the address on them looks like. */
    private static final String[] MEDIA_HOSTS = {
            "i.redd.it", "preview.redd.it", "v.redd.it", "i.imgur.com", "imgur.com",
            "redgifs.com", "gfycat.com", "giphy.com", "tenor.com", "media.giphy.com",
            "i.redditmedia.com", "redditmedia.com", "imgflip.com", "i.imgflip.com",
            "streamable.com", "media.tumblr.com", "youtube.com", "youtu.be",
    };

    /** What an address ends with when it is media, whatever host it is on. */
    private static final String[] MEDIA_ENDINGS = {
            ".jpg", ".jpeg", ".png", ".gif", ".gifv", ".webp", ".bmp", ".apng",
            ".mp4", ".webm", ".mov", ".m4v",
    };

    private static boolean said;

    /** How many calls are worth writing down before the point is made. */
    private static final int WORTH_SAYING = 200;

    private static int parsed;
    private static int gated;
    private static int asked;

    private InlineCommentMediaPatch() {}

    private static int drew;

    private static int seen;

    /**
     * Called with every address the link handling is given, before any of it is decided.
     *
     * <p>Said for each one because a capture showing only some of them cannot be told from a
     * capture where only some were handled, and which of the two it was has mattered twice.
     */
    public static void linkSeen(String link) {
        if (seen < WORTH_SAYING) {
            seen++;
            Logger.printInfo(() -> "inline comments: handling the link " + link);
        }
    }

    /** Set while a picture has just been put in, so the address is not written out after it. */
    private static final ThreadLocal<Boolean> justDrew = new ThreadLocal<>();

    /**
     * Draws the picture where the link stood, rather than letting Sync make a card of it.
     *
     * <p>Sync has two ways of putting a picture into text. Its giphy handling puts the picture
     * in by itself. Everything else goes into a list of cards, and a card is drawn as a small
     * picture with the address beside it — which is the rectangle, and is not what was asked
     * for. This puts it in the first way.
     *
     * @param card What Sync was about to do, which is make a card of it.
     * @param into The text being built.
     * @param link The address of the picture.
     * @return Whether Sync should still make its card.
     */
    /** What text stands for a picture that is drawn rather than named. */
    private static final String THE_PICTURE_ITSELF = "\uFFFC";

    /**
     * @return The text to write where the link stood.
     *
     * <p>Sync is about to add the link to the cards it draws under the text — a small picture
     * with the address beside it. Where the picture itself was asked for, no card is added and
     * the text written is the one character a picture stands in, which the spans below fill.
     *
     * <p>Adding the card as well is what drew each picture twice.
     */
    public static String cardOrPicture(nc.b cards, nc.b.a card, String link) {
        if (!wanted(link)) {
            return cards.c(card);
        }
        justDrew.set(Boolean.TRUE);
        Logger.printInfo(() -> "inline comments: no card for " + link
                + "; the picture stands where the address did");
        // Nothing at all: the picture has already been drawn over the address itself, and a
        // card would be the same picture again, small, underneath it.
        return "";
    }

    /**
     * @return What to draw that text with: the picture by itself, which is what Sync's own
     *         giphy handling draws one with, rather than the card's label.
     */
    public static Object[] spansFor(Object[] spans, String link, nc.a where) {
        if (!wanted(link)) {
            return spans;
        }
        try {
            int width = where == null ? 0 : where.e;
            if (width > 0) {
                // Remembered for Reddit's own giphy pictures, which are drawn by a path that is
                // told the address and nothing else.
                lastWidth = width;
            }
            if (width <= 0) {
                Logger.printInfo(() -> "inline comments: no width to draw " + link + " at");
                return spans;
            }
            Logger.printInfo(() -> "inline comments: drawing " + link + " " + width + " wide");
            // The span that draws the picture and nothing else, at the width the text has.
            // Its sibling draws a card: a small picture with the address beside it.
            return new Object[]{ new nb.c(link, width), new mb.d(link) };
        } catch (Throwable ex) {
            Logger.printInfo(() -> "inline comments: could not draw " + link + ": " + ex);
            return spans;
        }
    }

    private static boolean wanted(String link) {
        return isPatchIncluded() && link != null && !link.isEmpty()
                && inlineEverything() && isMedia(link);
    }

    /**
     * @param expanded Whether Sync writes the address out after a link.
     * @param handled  Whether the link was already drawn as something.
     * @return Whether to write the address out.
     */
    public static boolean stillShowTheLink(boolean expanded, boolean handled) {
        if (Boolean.TRUE.equals(justDrew.get())) {
            justDrew.remove();
            Logger.printDebug(() -> "inline comments: leaving the address off what was drawn");
            return false;
        }
        return expanded;
    }

    /** How wide the text was last drawn, for a path that is not told. */
    private static volatile int lastWidth;

    /**
     * The giphy picture being drawn.
     *
     * <p>Kept here rather than read again further down: Sync builds the address into a register
     * and then uses that same register for the character a picture stands in, so by the time
     * the spans are written the address is no longer there to read.
     */
    private static volatile String giphyLink;

    /** Called where Sync draws one of Reddit's own giphy pictures, which it has its own path for. */
    public static void giphy(String link) {
        giphyLink = link;
        Logger.printInfo(() -> "inline comments: a giphy picture, drawn from " + link);
    }

    /**
     * @return What to draw Reddit's own giphy pictures with.
     *
     * <p>They have a path of their own, which draws a card: a small picture with the address
     * beside it. Where the picture itself was asked for, it is drawn as the picture, at the
     * width the text was last drawn at — that path is told the address and nothing else.
     */
    public static Object[] giphySpans(Object[] spans) {
        String link = giphyLink;
        if (!isPatchIncluded() || !inlineEverything() || link == null || link.isEmpty()) {
            return spans;
        }
        int width = lastWidth;
        if (width <= 0) {
            try {
                width = (int) (app.morphe.extension.shared.Utils.getContext().getResources()
                        .getDisplayMetrics().widthPixels * 0.9f);
            } catch (Throwable ex) {
                Logger.printInfo(() -> "inline comments: no width for a giphy picture: " + ex);
                return spans;
            }
        }
        final int drawAt = width;
        try {
            Logger.printInfo(() -> "inline comments: drawing the giphy picture " + link
                    + " " + drawAt + " wide");
            return new Object[]{ new nb.c(link, drawAt), new mb.d(link) };
        } catch (Throwable ex) {
            Logger.printInfo(() -> "inline comments: could not draw " + link + ": " + ex);
            return spans;
        }
    }

    /**
     * Called wherever text with markup in it is turned into something drawable, before any of
     * it is looked at.
     *
     * <p>Nothing at all was heard from the link handling, and a capture that says nothing
     * cannot tell "the code declined" from "the code never ran". This says which.
     *
     * @param markup What is about to be turned into drawable text.
     */
    public static void drawing(String markup) {
        if (drew < WORTH_SAYING) {
            drew++;
            String head = markup == null ? "nothing"
                    : markup.substring(0, Math.min(120, markup.length())).replace('\n', ' ');
            Logger.printInfo(() -> "inline comments: drawing text (" + drew + "): " + head);
        }
    }

    /**
     * Called for every link drawn in a post or a comment, before anything is decided.
     *
     * <p>Said unconditionally for the first few: a capture with none of these in it means this
     * is not the code that draws the thing in question, rather than that it declined to.
     */
    public static void parsing() {
        if (parsed < WORTH_SAYING) {
            parsed++;
            Logger.printInfo(() -> "inline comments: a link is being drawn (" + parsed + ")");
        }
    }

    /**
     * Sync asks something about where the link is before it asks anything about the link. What
     * that is, is not named in the app; it is answered here so that a capture says whether it
     * is what stops a picture being drawn where it sits.
     *
     * @param sync What Sync decided about where this link is.
     */
    public static boolean here(boolean sync) {
        boolean on = inlineEverything();
        if (gated < WORTH_SAYING) {
            gated++;
            Logger.printInfo(() -> "inline comments: Sync says this is "
                    + (sync ? "" : "not ") + "somewhere to draw one, widening is "
                    + (on ? "on" : "off"));
        }
        return sync || on;
    }

    public static boolean isPatchIncluded() {
        // Overridden by patch.
        return false;
    }

    /** Whether the setting is on, so Sync's own gate can be opened alongside it. */
    public static boolean inlineEverything() {
        if (!isPatchIncluded()) {
            return false;
        }
        boolean on = SyncUpSettings.flag(INLINE_EVERYTHING, false);
        if (!said) {
            said = true;
            SyncUpSettings.report("inline comments", INLINE_EVERYTHING);
            Logger.printInfo(() -> "inline comments: widening is " + (on ? "on" : "off"));
        }
        return on;
    }

    /**
     * Sync will not draw anything inline unless its own inline-previews setting is on, and that
     * setting is about its own narrow list. Opening the gate here lets the widened test below be
     * the one that decides.
     *
     * @param sync What Sync's own setting says.
     */
    public static boolean orInlineEverything(boolean sync) {
        boolean on = inlineEverything();
        if (asked < WORTH_SAYING) {
            asked++;
            Logger.printInfo(() -> "inline comments: Sync's own setting is "
                    + (sync ? "on" : "off") + ", widening is " + (on ? "on" : "off"));
        }
        return sync || on;
    }

    /**
     * Whether this link is one to draw where it sits rather than to show a preview of the page
     * it is on.
     *
     * <p>Sync decides whether to preview a page before it decides anything about drawing a
     * picture, and a link it has already made a preview of is marked as dealt with and never
     * looked at again. A gif was therefore always a preview of the page it is on, whatever the
     * rest of this answered.
     */
    public static boolean claimsAsMedia(String link) {
        if (!isPatchIncluded() || link == null || link.isEmpty() || !inlineEverything()) {
            return false;
        }
        boolean media = isMedia(link);
        if (media) {
            Logger.printInfo(() -> "inline comments: claiming " + link
                    + " rather than letting it become a preview");
        }
        return media;
    }

    /**
     * @param sync     What Sync decided on its own.
     * @param link     The address in the comment.
     * @return Whether to draw the link as the media it points at.
     */
    public static boolean shouldInline(boolean sync, String link) {
        if (sync) {
            Logger.printDebug(() -> "inline comments: Sync already draws " + link);
            return true;
        }
        if (!inlineEverything()) {
            Logger.printDebug(() -> "inline comments: leaving " + link + " as a chip");
            return false;
        }
        if (link == null || link.isEmpty()) {
            return false;
        }

        boolean media = isMedia(link);
        Logger.printInfo(() -> "inline comments: " + link + " -> "
                + (media ? "drawing it inline" : "not media, left as a chip"));
        return media;
    }

    /** @return Whether the address points at something that can be drawn. */
    private static boolean isMedia(String link) {
        String lower = link.toLowerCase(Locale.ROOT);

        // Anything after the address itself is no help in telling what it points at.
        int query = lower.indexOf('?');
        String path = query < 0 ? lower : lower.substring(0, query);
        for (String ending : MEDIA_ENDINGS) {
            if (path.endsWith(ending)) {
                return true;
            }
        }

        for (String host : MEDIA_HOSTS) {
            // Matched against the whole address rather than a parsed host, since what arrives
            // here is sometimes written without a scheme.
            if (lower.contains("//" + host + "/") || lower.contains("." + host + "/")
                    || lower.contains("//" + host) && lower.endsWith(host)) {
                return true;
            }
        }
        return false;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/customclients/sync/syncforreddit/ui/comments/InlineCommentMediaPatch.kt
package app.morphe.patches.reddit.customclients.sync.syncforreddit.ui.comments

import app.morphe.patcher.Fingerprint
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.extensions.InstructionExtensions.instructions
import app.morphe.patcher.patch.PatchException
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.reddit.customclients.sync.SyncForRedditCompatible
import app.morphe.patches.reddit.customclients.sync.syncforreddit.extension.sharedExtensionPatch
import app.morphe.patches.reddit.customclients.sync.syncforreddit.http.interceptHttpRequests
import app.morphe.util.getReference
import app.morphe.util.returnEarly
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.instruction.FiveRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.reference.FieldReference
import com.android.tools.smali.dexlib2.iface.reference.MethodReference
import com.android.tools.smali.dexlib2.iface.reference.StringReference
import app.morphe.patcher.extensions.InstructionExtensions.replaceInstruction
import com.android.tools.smali.dexlib2.iface.reference.TypeReference

private const val EXTENSION_CLASS_DESCRIPTOR =
    "Lapp/morphe/extension/syncforreddit/ui/comments/InlineCommentMediaPatch;"

/** Reads the sizes Reddit sends beside the comments, so a picture is drawn right the first time. */
private const val SIZES_CLASS_DESCRIPTOR =
    "Lapp/morphe/extension/syncforreddit/ui/comments/CommentMediaSizes;"

private const val SHOULD_INLINE_METHOD = "shouldInline(ZLjava/lang/String;)Z"

private const val OPEN_THE_GATE_METHOD = "orInlineEverything(Z)Z"

private const val HERE_METHOD = "here(Z)Z"

private const val GIPHY_METHOD = "giphy(Ljava/lang/String;)V"

private const val GIPHY_SPANS_METHOD = "giphySpans([Ljava/lang/Object;)[Ljava/lang/Object;"

private const val CARD_OR_PICTURE_METHOD =
    "cardOrPicture(Lnc/b;Lnc/b\$a;Ljava/lang/String;)Ljava/lang/String;"

private const val HOW_WIDE_METHOD = "howWide(Lnc/a;)V"

private const val SHAPE_FOR_METHOD = "shapeFor(Lt3/a;Lnb/c;)Lt3/a;"

/** Where Sync builds the Glide request for every span it draws in a piece of text. */
private const val TEXT_PICTURE_LOADER = "Loc/c;"

/** What Glide is told to decode a picture with. */
private const val GLIDE_OPTIONS = "Lt3/a;"

/** The span that draws a picture and nothing else. Its sibling draws a card. */
private const val PICTURE_SPAN = "Lnb/c;"

/** Where Sync works out how wide a comment's text may be drawn. */
private const val COMMENT_TEXT_VIEW =
    "Lcom/laurencedawson/reddit_sync/ui/views/comments/CommentsHtmlTextView;"

private const val SPANS_FOR_METHOD =
    "spansFor([Ljava/lang/Object;Ljava/lang/String;Lnc/a;)[Ljava/lang/Object;"

private const val STILL_SHOW_METHOD = "stillShowTheLink(ZZ)Z"

/** Sync's own setting for writing the address out after a link it has drawn. */
private const val WRITES_THE_ADDRESS_OUT = "commentsLinksExpanded"

/**
 * What Sync asks giphy for when it draws one of Reddit's own giphy pictures: a hundred pixels
 * of it. That is a thumbnail, not the picture, which is why one drawn where it sits looked like
 * a chip beside the address rather than the gif itself.
 */
private const val A_THUMBNAIL_OF_IT = "/100.gif"

/** Big enough to read in a comment without fetching the whole of a very large gif. */
private const val ENOUGH_OF_IT = "/200.gif"

/** What Sync asks about where a link is, before it asks anything about the link itself. */
private const val WHERE_IT_IS = "Lnc/a;"

/** The span Sync draws a link as when it draws the picture rather than a chip naming it. */
private const val INLINE_SPAN = "Lnb/d;"

/** Sync's own setting, which only ever covered its own narrow list of addresses. */
private const val ITS_OWN_SETTING = "inlineImagePreviews"

/**
 * What Sync asks before it asks anything else about the address: whether it is an Imgur one.
 * Anything that is not, and is not already known to it, leaves here and never reaches the test
 * further down — which is why a gif hosted anywhere else was drawn as a chip however the rest of
 * it answered.
 */
private const val THE_IMGUR_TEST = "Lf8/a;"

@Suppress("unused")
val inlineCommentMediaPatch = bytecodePatch(
    name = "Show media in a comment where it sits",
    description = "Draws a link in a comment as the picture or video it points at rather than " +
            "as a chip naming where it goes.",
    default = true
) {
    dependsOn(sharedExtensionPatch, inlineCommentMediaSettingsPatch, interceptHttpRequests)

    compatibleWith(*SyncForRedditCompatible)

    execute {
        Fingerprint(
            definingClass = EXTENSION_CLASS_DESCRIPTOR,
            name = "isPatchIncluded",
        ).method.returnEarly(true)

        Fingerprint(
            definingClass = SIZES_CLASS_DESCRIPTOR,
            name = "isPatchIncluded",
        ).method.returnEarly(true)

        // How wide the text may be drawn. Sync takes the paddings off the width of the screen,
        // which at a reply to a reply comes back larger than the room actually there, so a
        // picture built to it is cut off on the right. Corrected where it is settled, which is
        // the one place the text view itself is to hand.
        Fingerprint(
            definingClass = COMMENT_TEXT_VIEW,
            name = "J",
            parameters = listOf("Lxa/d;", "Ljava/lang/String;"),
            returnType = "V",
        ).method.apply {
            val saysHowWide = instructions.indexOfFirst {
                it.opcode == Opcode.INVOKE_VIRTUAL &&
                    it.getReference<MethodReference>()?.let { asked ->
                        asked.definingClass == WHERE_IT_IS && asked.name == "d"
                    } == true
            }
            if (saysHowWide < 0) {
                throw PatchException("Sync no longer says how wide a comment's text may be")
            }
            val where = getInstruction<FiveRegisterInstruction>(saysHowWide).registerC
            addInstructions(
                saysHowWide + 1,
                "invoke-static { v$where }, $EXTENSION_CLASS_DESCRIPTOR->$HOW_WIDE_METHOD",
            )
        }

        // Where a link in a comment is turned into either a drawn picture or a chip. Anchored on
        // the span itself, since the decision around it is a long run of unnamed tests.
        Fingerprint(
            definingClass = "Lnc/d;",
            name = "e",
            parameters = listOf("Loc/c;", "Lnc/a;", "Lnc/b;"),
            returnType = "V",
        ).method.apply {
            val spanIndex = instructions.indexOfFirst {
                it.opcode == Opcode.NEW_INSTANCE &&
                        it.getReference<TypeReference>()?.type == INLINE_SPAN
            }
            if (spanIndex < 0) {
                throw PatchException("Nothing in a comment is drawn where it sits")
            }

            // The address is whatever the span is built out of, rather than a register guessed
            // at: the method reuses its own registers throughout.
            val builtIndex = (spanIndex until instructions.size).first { at ->
                val reference = getInstruction(at).getReference<com.android.tools.smali.dexlib2.iface.reference.MethodReference>()
                reference?.definingClass == INLINE_SPAN && reference.name == "<init>"
            }
            val linkRegister = getInstruction<FiveRegisterInstruction>(builtIndex).registerD

            // What guards the span is the last thing asked before it is built.
            val guardIndex = (spanIndex - 1 downTo 0).first {
                instructions.elementAt(it).opcode == Opcode.IF_EQZ
            }
            val decisionRegister = getInstruction<OneRegisterInstruction>(guardIndex).registerA

            // Where the card is finally written into the text. Chosen over the test that
            // guards it because a branch further up jumps to that test rather than through it,
            // so anything put in front of it is stepped over by every link but a tenor one.
            // Where the address itself is styled. The span that draws a picture stands in
            // place of the text it covers, so putting one here is what turns the address into
            // the picture rather than leaving both.
            instructions.withIndex()
                .filter { (_, instruction) ->
                    instruction.getReference<MethodReference>()?.let {
                        it.definingClass == "Lnc/d;" && it.name == "r"
                    } == true
                }
                .map { (at, _) -> at }
                .reversed()
                .forEach { at ->
                    val spans = getInstruction<FiveRegisterInstruction>(at).registerE
                    addInstructions(
                        at,
                        """
                        invoke-static       { v$spans, v$linkRegister, p1 }, $EXTENSION_CLASS_DESCRIPTOR->$SPANS_FOR_METHOD
                        move-result-object  v$spans
                        """
                    )
                }

            // And the card underneath it, which would be the same picture again, small.
            val addsIndex = (spanIndex until instructions.count()).first { at ->
                val called = getInstruction(at).getReference<MethodReference>()
                called?.definingClass == "Lnc/b;" && called.name == "c"
            }
            val adds = getInstruction<FiveRegisterInstruction>(addsIndex)
            replaceInstruction(
                addsIndex,
                "invoke-static { v${adds.registerC}, v${adds.registerD}, v$linkRegister }, " +
                    "$EXTENSION_CLASS_DESCRIPTOR->$CARD_OR_PICTURE_METHOD"
            )

            // Where the card is written into the text. What Sync puts there is a card — a
            // small picture with the address beside it — and what goes there instead is the
            // picture itself, drawn at the width the text has.
            val cardIndex = (spanIndex until instructions.count()).first { at ->
                val called = getInstruction(at).getReference<MethodReference>()
                called?.definingClass == "Loc/c;" && called.name == "c"
            }
            val spansRegister = getInstruction<FiveRegisterInstruction>(cardIndex).registerE

            addInstructions(
                cardIndex,
                """
                invoke-static       { v$spansRegister, v$linkRegister, p1 }, $EXTENSION_CLASS_DESCRIPTOR->$SPANS_FOR_METHOD
                move-result-object  v$spansRegister
                """
            )

            // Sync writes the address out after a link when that is turned on, which would
            // stand under the picture just put in.
            val writesIndex = instructions.indexOfFirst {
                it.opcode == Opcode.IGET_BOOLEAN &&
                        it.getReference<FieldReference>()?.name == WRITES_THE_ADDRESS_OUT
            }
            if (writesIndex >= 0) {
                val writes = getInstruction<OneRegisterInstruction>(writesIndex).registerA
                addInstructions(
                    writesIndex + 1,
                    """
                    invoke-static       { v$writes, v$writes }, $EXTENSION_CLASS_DESCRIPTOR->$STILL_SHOW_METHOD
                    move-result         v$writes
                    """
                )
            }

            // The first of the two, and the one that actually turned a gif away: anything not
            // on Imgur leaves here, long before the test above is reached.
            val imgurIndex = instructions.indexOfFirst {
                it.opcode == Opcode.INVOKE_STATIC &&
                        it.getReference<MethodReference>()?.definingClass == THE_IMGUR_TEST
            }
            if (imgurIndex < 0) {
                throw PatchException("Sync no longer asks whether a link in a comment is Imgur")
            }
            val imgurResult = imgurIndex + 1
            val imgurRegister = getInstruction<OneRegisterInstruction>(imgurResult).registerA

            addInstructions(
                imgurResult + 1,
                """
                invoke-static       { v$imgurRegister, v$linkRegister }, $EXTENSION_CLASS_DESCRIPTOR->$SHOULD_INLINE_METHOD
                move-result         v$imgurRegister
                """
            )

            // Sync asks its own setting first and gives up where it is off, so that has to be
            // answered too or the widened test above is never reached.
            val settingIndex = instructions.indexOfFirst {
                it.opcode == Opcode.IGET_BOOLEAN &&
                        it.getReference<FieldReference>()?.name == ITS_OWN_SETTING
            }
            if (settingIndex < 0) {
                throw PatchException("Sync no longer asks whether to draw a link in a comment")
            }
            val settingRegister = getInstruction<OneRegisterInstruction>(settingIndex).registerA

            addInstructions(
                settingIndex + 1,
                """
                invoke-static       { v$settingRegister }, $EXTENSION_CLASS_DESCRIPTOR->$OPEN_THE_GATE_METHOD
                move-result         v$settingRegister
                """
            )

            // Asked before either of those: something about where the link is, which is not
            // named in the app. Answered so that a capture says whether it is what stops a
            // picture being drawn where it sits, and opened where it is.
            val whereIndex = (settingIndex - 1 downTo 0).firstOrNull { at ->
                val instruction = instructions.elementAt(at)
                instruction.opcode == Opcode.IGET_BOOLEAN &&
                        instruction.getReference<FieldReference>()?.definingClass == WHERE_IT_IS
            }
            if (whereIndex != null) {
                val whereRegister = getInstruction<OneRegisterInstruction>(whereIndex).registerA
                addInstructions(
                    whereIndex + 1,
                    """
                    invoke-static       { v$whereRegister }, $EXTENSION_CLASS_DESCRIPTOR->$HERE_METHOD
                    move-result         v$whereRegister
                    """
                )
            }

        }

        // How a picture drawn in a comment is decoded. Sync asks for it in a square of the
        // span's width and rounds it by 8dp, and the span then stretches whatever comes back to
        // the size it was made with; the rounding being in the decoded picture's own pixels, the
        // corners came out that radius times however far it was stretched. Asked for at the size
        // it is drawn and scaled to exactly that before rounding, which is what Sync already does
        // for every other picture it draws in text.
        Fingerprint(
            definingClass = TEXT_PICTURE_LOADER,
            name = "o",
            parameters = listOf("Loc/b;"),
            returnType = "V",
        ).method.apply {
            // The only picture in this method rounded on its own, without being scaled first:
            // every other kind is given a list of two, so one transformation names this one.
            val rounds = instructions.withIndex().filter { (_, instruction) ->
                instruction.opcode == Opcode.INVOKE_VIRTUAL &&
                    instruction.getReference<MethodReference>()?.let { asked ->
                        asked.definingClass == GLIDE_OPTIONS && asked.name == "j0"
                    } == true
            }
            if (rounds.size != 1) {
                throw PatchException(
                    "Expected one picture rounded without being scaled, found ${rounds.size}",
                )
            }
            val at = rounds.single().index
            // The receiver is the request being built; the argument beside it is the rounding
            // itself, not the span, so the span is taken from where it was named just above.
            val request = getInstruction<FiveRegisterInstruction>(at).registerC
            val named = (at - 1 downTo 0).firstOrNull { step ->
                val instruction = instructions.elementAt(step)
                instruction.opcode == Opcode.CHECK_CAST &&
                    instruction.getReference<TypeReference>()?.type == PICTURE_SPAN
            } ?: throw PatchException("Nothing says which span the picture is being decoded for")
            val span = getInstruction<OneRegisterInstruction>(named).registerA
            replaceInstruction(
                at,
                "invoke-static { v$request, v$span }, " +
                    "$EXTENSION_CLASS_DESCRIPTOR->$SHAPE_FOR_METHOD",
            )
        }

        // Reddit's own giphy pictures have a path of their own, which asks for a hundred
        // pixels of the gif. Said so, and asked for enough of it to be worth drawing.
        mutableClassDefBy("Lnc/d;").methods
            .filter { it.name == "a" && it.parameters.size == 2 && it.implementation != null }
            .forEach { method ->
                val at = method.implementation!!.instructions.toList().indexOfFirst {
                    it.opcode == Opcode.CONST_STRING &&
                            it.getReference<StringReference>()?.string == A_THUMBNAIL_OF_IT
                }
                if (at < 0) {
                    return@forEach
                }
                val register = method.getInstruction<OneRegisterInstruction>(at).registerA
                method.replaceInstruction(at, "const-string v$register, \"$ENOUGH_OF_IT\"")

                // Said once the address has been built out of it.
                val built = (at until method.implementation!!.instructions.count()).first { on ->
                    method.getInstruction(on).opcode == Opcode.MOVE_RESULT_OBJECT &&
                            on > at + 1
                }
                val link = method.getInstruction<OneRegisterInstruction>(built).registerA

                // Drawn as the picture rather than as a card, the same as every other one.
                val writes = method.implementation!!.instructions.toList().indexOfFirst {
                    it.getReference<MethodReference>()?.let { called ->
                        called.definingClass == "Loc/c;" && called.name == "c"
                    } == true
                }
                if (writes >= 0) {
                    val spans = method.getInstruction<FiveRegisterInstruction>(writes).registerE
                    method.addInstructions(
                        writes,
                        """
                        invoke-static       { v$spans }, $EXTENSION_CLASS_DESCRIPTOR->$GIPHY_SPANS_METHOD
                        move-result-object  v$spans
                        """
                    )
                }

                method.addInstructions(
                    built + 1,
                    "invoke-static { v$link }, $EXTENSION_CLASS_DESCRIPTOR->$GIPHY_METHOD"
                )
            }

    }
}
